src/app-layer-register.c

```diff
@@ -0,0 +1,176 @@
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Pierre Chifflier <chifflier@wzdftpd.net>
+ *
+ * Parser registration functions.
+ */
+
+#include "suricata-common.h"
+#include "stream.h"
+#include "conf.h"
+
+#include "app-layer-detect-proto.h"
+#include "app-layer-parser.h"
+
+#include "app-layer-register.h"
+
+AppProto AppLayerRegisterParser(const struct AppLayerParser *p)
+{
+    AppProto alproto;
+    const char *ip_proto_str = NULL;
+
+    if (p == NULL)
+        return ALPROTO_FAILED;
+
+    alproto = StringToAppProto(p->name);
+    if (alproto == ALPROTO_UNKNOWN || alproto == ALPROTO_FAILED)
+        return alproto;
+
+    switch (p->ip_proto) {
+        case IPPROTO_TCP:
+            ip_proto_str = "tcp";
+            break;
+        case IPPROTO_UDP:
+            ip_proto_str = "udp";
+            break;
+        default:
+            SCLogDebug("Unknown or unsupported ip_proto field in parser");
+            return ALPROTO_UNKNOWN;
+    };
+
+    /* Check if protocol detection is enabled. If it does not exist in
+     * the configuration file then it will be enabled by default. */
+    if (AppLayerProtoDetectConfProtoDetectionEnabled(ip_proto_str, p->name)) {
+
+        SCLogDebug("%s %s protocol detection enabled.", ip_proto_str, p->name);
+
+        AppLayerProtoDetectRegisterProtocol(alproto, p->name);
+
+        if (RunmodeIsUnittests()) {
+
+            SCLogDebug("Unittest mode, registering default configuration.");
+            AppLayerProtoDetectPPRegister(p->ip_proto, p->default_port,
+                alproto, p->min_depth, p->max_depth, STREAM_TOSERVER,
+                p->ProbeTS, p->ProbeTC);
+
+        }
+        else {
+
+            if (!AppLayerProtoDetectPPParseConfPorts(ip_proto_str, p->ip_proto,
+                    p->name, alproto, p->min_depth, p->max_depth,
+                    p->ProbeTS, p->ProbeTC)) {
+                SCLogDebug("No %s app-layer configuration, enabling %s"
+                    " detection %s detection on port %s.",
+                    p->name, p->name, ip_proto_str, p->default_port);
+                AppLayerProtoDetectPPRegister(p->ip_proto,
+                    p->default_port, alproto,
+                    p->min_depth, p->max_depth, STREAM_TOSERVER,
+                    p->ProbeTS, p->ProbeTC);
+            }
+
+        }
+    }
+    else {
+        SCLogDebug("Protocol detecter and parser disabled for %s.", p->name);
+        return alproto;
+    }
+
```

Comment:
Would it be reasonable to refactor this into 2 functions. So parsers that don't follow default logic for being enabled could still use the registration bit?

Reply:
Sure, done for the next PR
This puts back some code when the parser registers itself (to check the config), but allows more flexibility: each protocol can choose the default behavior, etc.